src/python/pants/backend/docker/rules_context.py

```diff
@@ -0,0 +1,100 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import logging
+from dataclasses import dataclass
+from itertools import chain
+from typing import Tuple
+
+from pants.backend.docker.target_types import DockerImageSources
+from pants.core.goals.package import BuiltPackage, PackageFieldSet
+from pants.core.target_types import FilesSources, ResourcesSources
+from pants.core.util_rules.source_files import SourceFiles, SourceFilesRequest
+from pants.engine.addresses import Address
+from pants.engine.fs import AddPrefix, Digest, MergeDigests
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.engine.target import (
+    Dependencies,
+    DependenciesRequest,
+    FieldSetsPerTarget,
+    FieldSetsPerTargetRequest,
+    Sources,
+    Target,
+    Targets,
+)
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class DockerBuildContext:
+    digest: Digest
+
+
+@dataclass(frozen=True)
+class DockerBuildContextRequest:
+    address: Address
+    context_root: str
+    targets: Tuple[Target, ...]
+
+
+@rule
+async def create_docker_build_context(request: DockerBuildContextRequest) -> DockerBuildContext:
+    # Get all sources, going into the build context.
```

Comment:
It looks like you're doing the expansion to transitive closure outside this rule? Maybe it should be inside this rule, to make it easier to grok? Unless there is a use-case for building a context from a smaller set of targets than the entire closure?

Reply:
It was to make the test slightly easier, to skip the need to setup a full env with deps and all. Easy enough to move in if it's not worth it 😄

Reply:
I'll postpone changing it until we start working on the tests..